java-manta-client/src/main/java/com/joyent/manta/client/multipart/EncryptedMultipartManager.java

```diff
@@ -277,9 +285,15 @@ public ByteArrayOutputStream call(final long uploadedBytes) throws IOException {
                             }
                         }
                     });
+            final MantaMultipartUploadPart part = wrapped.uploadPart(upload.getWrapped(), partNumber, entity);
             encryptionState.setLastPartNumber(partNumber);
-            return wrapped.uploadPart(upload.getWrapped(), partNumber, entity);
+            return part;
         } finally {
+            if (encryptionState.getLastPartNumber() != partNumber && snapshot != null) {
```

Comment:
Using an inner `try` now but yes I'm hesitant to classify exceptions by their likelihood of having affected `Cipher` state an only rewind in certain cases, especially given the narrow scope of the inner `try` block. It seems much more straightforward to _always_ reset encryption state if we don't make it to the `setLastPartNumber` invocation.